Update 3.4.1 breaks building Apache Thrift (permission issue?) on Ubuntu 18.04

This issue has been reported originally as https://github.com/dotnet/core/issues/4719, and @dagood has done a great job to isolate a potential cause. I'll summarize the original issue and dagood's findings below, in case this is _indeed_ an msbuild issue. If its not related to msbuild, please forgive.

@emmenlau's original report:
I've been using dotnet on Ubuntu Linux 18.04 x86_64 (installed as in https://docs.microsoft.com/en-us/dotnet/core/install/linux-package-manager-ubuntu-1804) for almost a year successfully to build https://github.com/apache/thrift. Yesterday I've installed system updates (which may have included an update of `dotnet`). Usually I install updates at least once every one or two months.

Since today, the build of Thrift fails with error:
```
[...]
cd /home/user/thrift/lib/netstd/Thrift && /usr/bin/dotnet msbuild Thrift.csproj /nologo /restore /target:Rebuild /property:Configuration=Debug /property:OutDir=/home/user/Debug/thrift/lib/netstd/Thrift/bin /property:BaseIntermediateOutputPath=/home/user/Debug/thrift/lib/netstd/Thrift/obj/
FAILED: lib/netstd/Thrift/bin/Thrift.dll 
cd /home/user/thrift/lib/netstd/Thrift && /usr/bin/dotnet msbuild Thrift.csproj /nologo /restore /target:Rebuild /property:Configuration=Debug /property:OutDir=/home/user/Debug/thrift/lib/netstd/Thrift/bin /property:BaseIntermediateOutputPath=/home/user/Debug/thrift/lib/netstd/Thrift/obj/
MSBUILD : error MSB1025: An internal failure occurred while running MSBuild.
System.ComponentModel.Win32Exception (13): Permission denied
   at System.Diagnostics.Process.set_PriorityClassCore(ProcessPriorityClass value)
   at System.Diagnostics.Process.set_PriorityClass(ProcessPriorityClass value)
   at Microsoft.Build.CommandLine.MSBuildApp.Execute(String[] commandLine)
Unhandled exception. System.ComponentModel.Win32Exception (13): Permission denied
   at System.Diagnostics.Process.set_PriorityClassCore(ProcessPriorityClass value)
   at System.Diagnostics.Process.set_PriorityClass(ProcessPriorityClass value)
   at Microsoft.Build.CommandLine.MSBuildApp.Execute(String[] commandLine)
   at Microsoft.Build.CommandLine.MSBuildApp.Main(String[] args)
```

I've tried to isolate the cause and could find so far:
 - No useful tips or workarounds with Google worked for me, or at least there was nothing that I could understand or map to this problem
 - When running `dotnet` with sudo permissions, the build works. This is not a usable workaround for me because it breaks my build environment in various places
 - My user account has read permissions on all files in `/usr/share/dotnet`. This was tested with reading all files with `grep -rn "Win32Exception" /usr/share/dotnet/ > /dev/zero` which reports no errors
 - Trying to upgrade `dotnet` lists no newer packages available as of today
 - Downgrading dotnet to the previous latest version 3.1.3 solves the problem and the build works. I've downgraded with:
```
sudo apt-get install dotnet-apphost-pack-3.1=3.1.3-1 dotnet-host=3.1.3-1 \
    dotnet-hostfxr-3.1=3.1.3-1 dotnet-runtime-3.1=3.1.3-1 dotnet-sdk-3.1=3.1.201-1 \
    aspnetcore-runtime-3.1=3.1.3-1 aspnetcore-targeting-pack-3.1=3.1.3-1 \
    dotnet-runtime-deps-3.1=3.1.3-1
```

This is an acceptable workaround for me, but sooner or later a fix would be appreciated.

@emmenlau then adds: One thing that might be related: My default process priority on Linux is not zero (as one would typically assume on Linux). I've set it slightly lower (higher integers give lower priority) to `5`. This gives better overall user experience on a multi-user machine.

@dagood then adds:
That stack trace goes (well, appears to go) here in particular:

https://github.com/microsoft/msbuild/blob/7869acfe0be8e983da5de514b896cca6f38fd645/src/MSBuild/XMake.cs#L628-L631

The priority values per enum value seem to be here (not sure where in docs):

https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/Common/src/Interop/Unix/System.Native/Interop.GetSetPriority.cs#L57-L58

So it does appear to be relying on that exact assumption you're talking about for default priority. As far as I can tell (I'm not familiar with this change, just found this issue interesting) it's trying to raise it back to "normal" 0. I'd suggest raising this issue in the MSBuild repo.

--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -627,7 +627,7 @@ namespace Microsoft.Build.CommandLine
 
                     // Honor the low priority flag, we place our selves below normal
                     // priority and let sub processes inherit that priority.
-                    ProcessPriorityClass priority = lowPriority ? ProcessPriorityClass.BelowNormal : ProcessPriorityClass.Normal;
+                    ProcessPriorityClass priority = lowPriority ? ProcessPriorityClass.BelowNormal : Process.GetCurrentProcess().PriorityClass;
                     Process.GetCurrentProcess().PriorityClass = priority;
 
                     DateTime t1 = DateTime.Now;

--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -625,8 +625,9 @@ namespace Microsoft.Build.CommandLine
                         Environment.SetEnvironmentVariable("MSBUILDLOADALLFILESASWRITEABLE", "1");
                     }
 
-                    // Honor the low priority flag, we place our selves below normal
-                    // priority and let sub processes inherit that priority.
+                    // Honor the low priority flag, we place our selves below normal priority and let sub processes inherit
+                    // that priority. Idle priority would prevent the build from proceeding as the user does normal actions.
+                    // We avoid increasing priority because that causes failures on mac/linux.
                     if (lowPriority && Process.GetCurrentProcess().PriorityClass != ProcessPriorityClass.Idle)
                     {
                         Process.GetCurrentProcess().PriorityClass = ProcessPriorityClass.BelowNormal;

--- a/src/MSBuild/XMake.cs
+++ b/src/MSBuild/XMake.cs
@@ -627,8 +627,10 @@ namespace Microsoft.Build.CommandLine
 
                     // Honor the low priority flag, we place our selves below normal
                     // priority and let sub processes inherit that priority.
-                    ProcessPriorityClass priority = lowPriority ? ProcessPriorityClass.BelowNormal : Process.GetCurrentProcess().PriorityClass;
-                    Process.GetCurrentProcess().PriorityClass = priority;
+                    if (lowPriority && Process.GetCurrentProcess().PriorityClass != ProcessPriorityClass.Idle)
+                    {
+                        Process.GetCurrentProcess().PriorityClass = ProcessPriorityClass.BelowNormal;
+                    }
 
                     DateTime t1 = DateTime.Now;
 

